tighter Slurm integration, documentation, bug fix

duylethanh opened the pull request:
This PR closes #92 and contains the following:

- query Slurm when encountering experiments submitted to or started by Slurm (details below)
- add ``Status.BROKEN`` for experiments that were launched through Slurm and have an unexpected Slurm status
- add ``simex experiments kill`` command (calls ``scancel`` internally)
      - ~``--submitted`` selects all experiments submitted to Slurm~
      - ~``--started`` selects all experiments started by Slurm~
      - uses the same run selection as in purge, launch etc.
      - by default: selects submitted and started experiments
- add documentation on Slurm launcher
- fix bug where the status cache would be written multiple times when using multiple launchers
- fix bug in run ``cli_selects_run()``, where it was not possible to select all runs that failed or were unfinished

Some details regarding the first bullet point:

When experiments are started by the Slurm launcher, the slurm jobid will be stored in the ``.run`` file. When simexpal determines ``submitted`` or ``started`` for a run that was started through Slurm and Slurm has not yet been queried (``Config.slurm_queried == False``), simexpal will query Slurm via ``squeue``. It will then parse the status of all listed jobs to avoid multiple ``squeue``commands and save them to ``Config.slurm_queried_jobs`` (key: Slurm jobid, value: ``simexpal.Status``) and set ``Config.slurm_queried = True``. Slurm statuses will be resolved as follows:
- Slurm status: ``PD`` -> simexpal: ``submitted``
- ``R``, ``CG`` -> ``started``
- any other (unexpected) status -> ``broken``

The parsed output can then be accesses via ~``Config.pop_slurm_queried_jobs(jobid)``~ ``Config.get_slurm_job_status(jobid)``. When a ``jobid`` that was not listed in ``squeue`` is provided, ``failed`` will be returned.

avdgrinten commented on simexpal/base.py:
```diff
@@ -419,6 +423,46 @@ def export_successful_experiments(self):
+			if len(self.slurm_queried_jobs) > 0:
+				self.slurm_queried = True
+
+	def pop_slurm_queried_jobs(self, jobid):
+		# Returns 'failed' for jobs not listed in self.slurm_queried_jobs.
+		# This only holds for jobs that would receive the status
+		# 'submitted' or 'started' in Run._update_status_cache_dict().
+		status = self.slurm_queried_jobs.pop(jobid, Status.FAILED)
+		if len(self.slurm_queried_jobs) == 0:
+			self.slurm_queried = False
+
+		return status
```
This function is a bit strange. Why do we pop the status instead of just returning it?

duylethanh replied:
Hm, I think I wanted to track whether the status of a job was already queried or not, but in hindsight it is probably not necessary

duylethanh replied:
done

avdgrinten commented on scripts/simex:
```diff
@@ -769,6 +770,51 @@ experiments_print_parser = experiments_subcmds.add_parser('print',
+
+experiments_kill_parser = experiments_subcmds.add_parser('kill')
+experiments_kill_parser.set_defaults(cmd=do_experiments_kill)
+experiments_kill_parser.add_argument('--submitted', action='store_true', help='Kill pending Slurm jobs')
+experiments_kill_parser.add_argument('--started', action='store_true', help='Kill running Slurm jobs')
```
Why do we use a different way to select runs in this function (compared to our usual options as in launch/purge/etc)?

duylethanh replied:
It only makes sense to kill experiments that were ``started`` or ``submitted``. But I can allow the usual options and then filter out every status that is not in ``['started', 'submitted']`` so the run selection interface stays consistent

avdgrinten replied:
Yes, that sounds reasonable.

duylethanh replied:
> It only makes sense to kill experiments that were `started` or `submitted`. But I can allow the usual options and then filter out every status that is not in `['started', 'submitted']` so the run selection interface stays consistent

done

duylethanh commented on simexpal/base.py:
```diff
@@ -1230,6 +1230,15 @@ def internal_name(self):
+	@property
+	def slurm_jobid(self):
+		try:
+			with open(self.aux_file_path('run'), 'r') as f:
+				data = yaml.load(f, Loader=YmlLoader)
+			return data.get('slurm_jobid', None)
```
Ah, ``yaml.load()`` will return ``None``. I will catch this case and return ``None`` for the ``slurm_jobid``.

duylethanh replied:
done

duylethanh commented on simexpal/launch/slurm.py:
```diff
@@ -103,10 +103,18 @@ def substitute(p):
 
-		for run in locked:
-			common.create_run_file(run)
+		jobid = return_msg.split()[3]
```
done

duylethanh commented:
I updated the PR and the description at the top.
Apart from the review comments, I adapted the documentation for ``simex e kill`` and fixed a bug in ``cli_selects_run()``, where it was not possible to select all runs that failed or were unfinished

Ready for review.

avdgrinten reviewed (approved):
LGTM.
